sentry-android-core/src/main/java/io/sentry/android/core/ViewHierarchyEventProcessor.java

```diff
@@ -0,0 +1,137 @@
+package io.sentry.android.core;
+
+import android.app.Activity;
+import android.view.View;
+import android.view.ViewGroup;
+import io.sentry.Attachment;
+import io.sentry.EventProcessor;
+import io.sentry.Hint;
+import io.sentry.SentryEvent;
+import io.sentry.SentryLevel;
+import io.sentry.android.core.internal.gestures.ViewUtils;
+import io.sentry.protocol.ViewHierarchy;
+import io.sentry.protocol.ViewHierarchyNode;
+import io.sentry.util.Objects;
+import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/** ViewHierarchyEventProcessor responsible for taking a snapshot of the current view hierarchy. */
+@ApiStatus.Internal
+public final class ViewHierarchyEventProcessor implements EventProcessor {
+
+  private final @NotNull SentryAndroidOptions options;
+
+  public ViewHierarchyEventProcessor(final @NotNull SentryAndroidOptions options) {
+    this.options = Objects.requireNonNull(options, "SentryAndroidOptions is required");
+  }
+
+  @Override
+  public @NotNull SentryEvent process(final @NotNull SentryEvent event, @NotNull Hint hint) {
+    if (!event.isErrored()) {
+      return event;
+    }
+
+    if (!options.isAttachViewHierarchy()) {
+      this.options.getLogger().log(SentryLevel.DEBUG, "attachViewHierarchy is disabled.");
+      return event;
+    }
+
+    final @Nullable Activity activity = CurrentActivityHolder.getInstance().getActivity();
+    if (activity == null) {
+      return event;
+    }
+
+    try {
+      final @NotNull ViewHierarchy viewHierarchy = snapshotViewHierarchy(activity);
+      @SuppressWarnings("CharsetObjectCanBeUsed")
+      final Charset UTF8 = Charset.forName("UTF-8");
+
+      try (final ByteArrayOutputStream stream = new ByteArrayOutputStream();
+          final Writer writer = new BufferedWriter(new OutputStreamWriter(stream, UTF8))) {
+
+        options.getSerializer().serialize(viewHierarchy, writer);
```

Comment:
This can be done out of the main thread, See this comment https://github.com/getsentry/sentry-java/pull/2440#discussion_r1055409486

Reply:
M: The charset can be created as a static field. https://github.com/getsentry/sentry-java/blob/f55020d90a282dc59a50203aead122e4d9e2c694/sentry/src/main/java/io/sentry/cache/CacheStrategy.java#L36-L37
So you don't need to create a new one every time.

Reply:
L: you can also fix it here -> https://github.com/getsentry/sentry-java/blob/f55020d90a282dc59a50203aead122e4d9e2c694/sentry/src/main/java/io/sentry/internal/modules/ModulesLoader.java#L42

Reply:
I moved the serialization to a background thread now and added a simple time guard (1s) to ensure it doesn't take too long - because otherwise it could potentially trigger an ANR, killing the app and losing the whole crash event.
The snapshotting of the view hierarchy itself is still done on the main thread, as one shouldn't touch UI widgets from a non-ui thread.

Reply:
I'd do it differently, see how this example is done:
`final Attachment image = new Attachment(imageFile.getAbsolutePath(), "sentry.png", "image/png");`
The file is read async when its being sent, so you can do the same by doing:
`new Attachment(imageFile.getAbsolutePath())`, you just need to do something similar to `SentryEnvelopeItem`, which has the `CachedItem` callback, so you do the serialization there.
Maybe pair up with @romtsn about that, or we can jump on a call and I show you around, also the Dart PR does exactly that, you can also take a look in there.
The current approach has the overhead of one more thread and still makes it a blocking call.

Reply:
Ah alright, yeah that's roughly what I initially [proposed here](https://github.com/getsentry/sentry-java/pull/2440#discussion_r1055320512) to postpone serialization. I've added a `bytesFactory` field to the `Attachment` class now, which gets consumed on demand in `SentryEnvelopeItem / CachedItem`.